// @mastra/core/dist/storage/constants.cjs
'use strict';

var chunkSU5APAM6_cjs = require('../chunk-SU5APAM6.cjs');



Object.defineProperty(exports, "AGENTS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.AGENTS_SCHEMA; }
});
Object.defineProperty(exports, "AGENT_VERSIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.AGENT_VERSIONS_SCHEMA; }
});
Object.defineProperty(exports, "DATASETS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.DATASETS_SCHEMA; }
});
Object.defineProperty(exports, "DATASET_ITEMS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.DATASET_ITEMS_SCHEMA; }
});
Object.defineProperty(exports, "DATASET_VERSIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.DATASET_VERSIONS_SCHEMA; }
});
Object.defineProperty(exports, "EXPERIMENTS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.EXPERIMENTS_SCHEMA; }
});
Object.defineProperty(exports, "EXPERIMENT_RESULTS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.EXPERIMENT_RESULTS_SCHEMA; }
});
Object.defineProperty(exports, "MCP_CLIENTS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.MCP_CLIENTS_SCHEMA; }
});
Object.defineProperty(exports, "MCP_CLIENT_VERSIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.MCP_CLIENT_VERSIONS_SCHEMA; }
});
Object.defineProperty(exports, "OBSERVATIONAL_MEMORY_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.OBSERVATIONAL_MEMORY_SCHEMA; }
});
Object.defineProperty(exports, "OBSERVATIONAL_MEMORY_TABLE_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.OBSERVATIONAL_MEMORY_TABLE_SCHEMA; }
});
Object.defineProperty(exports, "OLD_SPAN_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.OLD_SPAN_SCHEMA; }
});
Object.defineProperty(exports, "PROMPT_BLOCKS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.PROMPT_BLOCKS_SCHEMA; }
});
Object.defineProperty(exports, "PROMPT_BLOCK_VERSIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.PROMPT_BLOCK_VERSIONS_SCHEMA; }
});
Object.defineProperty(exports, "SCORERS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.SCORERS_SCHEMA; }
});
Object.defineProperty(exports, "SCORER_DEFINITIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.SCORER_DEFINITIONS_SCHEMA; }
});
Object.defineProperty(exports, "SCORER_DEFINITION_VERSIONS_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.SCORER_DEFINITION_VERSIONS_SCHEMA; }
});
Object.defineProperty(exports, "SPAN_SCHEMA", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.SPAN_SCHEMA; }
});
Object.defineProperty(exports, "TABLE_AGENTS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_AGENTS; }
});
Object.defineProperty(exports, "TABLE_AGENT_VERSIONS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_AGENT_VERSIONS; }
});
Object.defineProperty(exports, "TABLE_CONFIGS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_CONFIGS; }
});
Object.defineProperty(exports, "TABLE_DATASETS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_DATASETS; }
});
Object.defineProperty(exports, "TABLE_DATASET_ITEMS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_DATASET_ITEMS; }
});
Object.defineProperty(exports, "TABLE_DATASET_VERSIONS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_DATASET_VERSIONS; }
});
Object.defineProperty(exports, "TABLE_EXPERIMENTS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_EXPERIMENTS; }
});
Object.defineProperty(exports, "TABLE_EXPERIMENT_RESULTS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_EXPERIMENT_RESULTS; }
});
Object.defineProperty(exports, "TABLE_MCP_CLIENTS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_MCP_CLIENTS; }
});
Object.defineProperty(exports, "TABLE_MCP_CLIENT_VERSIONS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_MCP_CLIENT_VERSIONS; }
});
Object.defineProperty(exports, "TABLE_MESSAGES", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_MESSAGES; }
});
Object.defineProperty(exports, "TABLE_OBSERVATIONAL_MEMORY", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_OBSERVATIONAL_MEMORY; }
});
Object.defineProperty(exports, "TABLE_PROMPT_BLOCKS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_PROMPT_BLOCKS; }
});
Object.defineProperty(exports, "TABLE_PROMPT_BLOCK_VERSIONS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_PROMPT_BLOCK_VERSIONS; }
});
Object.defineProperty(exports, "TABLE_RESOURCES", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_RESOURCES; }
});
Object.defineProperty(exports, "TABLE_SCHEMAS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_SCHEMAS; }
});
Object.defineProperty(exports, "TABLE_SCORERS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_SCORERS; }
});
Object.defineProperty(exports, "TABLE_SCORER_DEFINITIONS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_SCORER_DEFINITIONS; }
});
Object.defineProperty(exports, "TABLE_SCORER_DEFINITION_VERSIONS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_SCORER_DEFINITION_VERSIONS; }
});
Object.defineProperty(exports, "TABLE_SPANS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_SPANS; }
});
Object.defineProperty(exports, "TABLE_THREADS", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_THREADS; }
});
Object.defineProperty(exports, "TABLE_TRACES", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_TRACES; }
});
Object.defineProperty(exports, "TABLE_WORKFLOW_SNAPSHOT", {
  enumerable: true,
  get: function () { return chunkSU5APAM6_cjs.TABLE_WORKFLOW_SNAPSHOT; }
});
